Arnold : Multiple Outputs Per Driver

danieldresser-ie opened the pull request:
( Recreated from #6207 but with the branch on GafferHQ ).

This allows rendering multi-layer exrs directly from Arnold.

It took a fair bit of reorganizing to fit this into the current structure, but I'm feeling reasonably good about how it ended up - it should be in a good place for reviewing at least. It probably could use more test coverage before we merge it ( perhaps less for the new feature being added, and more for some of the existing functionality, to make sure that nothing has been broken that is currently in use ).

The only known breakage that happened in the tests were that there were some tests that used the same filename for multiple outputs - these failed with an error about these outputs not being possible to merge, until they were given distinct names, which seems to reflect the intent of the tests. Hopefully this wouldn't be an issue in production scenes, where people are actually using the files ( the only reason why the file being overwritten wasn't previously an issue is because the test in question was just testing the conversion to an Arnold scene, not an actual render ).

The one case I can think of that could affect production is that previously when using ieDisplay, you could use an identical fileName for multiple outputs with different parameters, since the fileName wasn't important to an interactive display. If users were doing that, they could an error message about the outputs not being valid to merge, and need to change the output name. But hopefully if they've been creating displays based on Gaffer's templates, this won't be an issue.

danieldresser-ie commented:
I was unable to reproduce the test failures locally. On a retry, everything succeeded in CI as well.

I'm still feeling rather concerned by one of the failures - this feels exactly like the sort of test that could reveal an issue where the new code does something wrong in a corner case:

`testEditLightGroups (GafferArnoldTest.InteractiveArnoldRenderTest.InteractiveArnoldRenderTest) ... gaffer: include/Gaffer/Signals.inl:271: Gaffer::Signals::Signal<Result(Args ...), Combiner>::Slot::CallScope::CallScope(Gaffer::Signals::Signal<Result(Args ...), Combiner>::Slot&) [with Result = void; Args = {GafferImage::GafferDisplayDriver*}; Combiner = Gaffer::Signals::DefaultCombiner<void>]: Assertion `slot.previous' failed.
WARNING | signal caught: SIGABRT -- Abnormal process abort`

Have you ever seen that before? I should probably try running the tests repeatedly locally - if it's a random thing, maybe I can reproduce a failure.

johnhaddon commented:
> Have you ever seen that before? 

I have, and I don't think it's related to this PR. I think it's down to Signal not being thread-safe, and something about the way we use signals in the Display node. I haven't found the time to look into it more deeply yet, partly because there's no sign of this causing problems in the wild.

johnhaddon reviewed (commented):
Thanks for taking this one on Daniel - it's great that we now have matching multi-layer output support between RenderMan, 3Delight and Arnold.

All looks pretty good to me - just a couple of comments about improving clarity in one spot, and moving the `updateInteractively` logic back to ArnoldOutput rather than having it in ArnoldDriver.

> It probably could use more test coverage before we merge it

You noted lack of coverage for the imager code path, so that one seems worth doing. And maybe that will do for now, or did you have other stuff in mind?

Since we're beginning to think about 1.6, and this is a little bit riskier than average, I suggest we retarget it to `main`. That also means we'll get to do plenty of incidental testing before it sees a release. We also need to update Changes.md - perhaps with the following :

- Arnold : Added multi-layer EXR support. All outputs with the same filename are now written to the same file via a single output driver.

Cheers...
John

danieldresser-ie commented on src/IECoreArnold/Renderer.cpp:
```diff
@@ -440,37 +440,35 @@ class ArnoldRendererBase : public IECoreScenePreview::Renderer
 
-class ArnoldOutput : public IECore::RefCounted
-{
+const IECore::InternedString g_cameraOverrideInternedString( "cameraOverride" );
```
I've gone ahead and done this. I agree it's a bit weird how it currently works ( although philosophically, it would make sense if it was a driver parameter, and then we wouldn't have to call AiRender multiple times with different cameras set ). I'm not sure how much the __ prefix in the string definition helps clarify things when you're looking at where it's used ... maybe it would be better to use a more verbose name? Though there are comments both where it's set and where it's read, so maybe it's OK?

johnhaddon replied:
I think my confusion came from seeing `cameraOverride` before the code using it, and knowing that `camera` was the name we used before - later I realised it was an accidental rename of the user-facing parameter, but an internal thing. I think __ does help clarify that.

danieldresser-ie commented on src/IECoreArnold/Renderer.cpp:
```diff
@@ -495,57 +493,52 @@ class ArnoldOutput : public IECore::RefCounted
-			std::vector<std::string> &customAttributes = customAttributesData->writable();
-			for( IECore::CompoundDataMap::const_iterator it = output->parameters().begin(), eIt = output->parameters().end(); it != eIt; ++it )
+			const IECore::BoolData *updateInteractivelyData = parameters->member<IECore::BoolData>( g_updateInteractivelyInternedString );
+			m_updateInteractively = updateInteractivelyData ? updateInteractivelyData->readable() : false;
```
I'd completely forgotten why I did this, but after testing: I suspect I did this because the ieDisplay driver doesn't work well with mixture of interactive and non-interactive outputs. When it receives an interactive update that doesn't include some layers, it deletes those layers, and they never seem to come back. If we're happy considering that as a bug in ieDisplay that we would need to fix in a hypothetical future where we send multiple AOVs through a single ieDisplay, then I can go ahead and switch to just directly using the per-output interactive flag ( it definitely does simplify the code ).

johnhaddon replied:
Yeah, I would call that an ieDisplay bug/limitation that we should fix later. Better to do the logical thing in ArnoldOutput now I think.

danieldresser-ie replied:
Done.

danieldresser-ie commented on src/IECoreArnold/Renderer.cpp:
```diff
@@ -3601,9 +3741,10 @@ class ArnoldGlobals
 					}
 				}
-				for( const auto &output : m_outputs )
+				// \todo - this didn't trip any tests when disabled ... suggests need for better test coverage here
```
Upon investigation, maybe the test coverage isn't as bad as I feared. There is actually some test coverage for imagers. 

What was actually happening is that previously there were two times when imagers would be updated on a driver:
A) when an output is updated, updating it's corresponding driver
B) when an imager option is set

In the new code, drivers are not updated dynamically ( I had considered this, but it gets quite messy - you don't know which outputs are sharing a driver until you get a render() call and know that all outputs are up to date ). For simplicity's sake, I just update all drivers for the current set of outputs when render() is called. I based this on A), and include correctly setting up the imager. Since this guarantees that the imager is always correct before rendering, there is no longer any need to call updateImager when the option is set. This resulted in my confusing observation that no tests fail when the updateImager call is removed from option().

Now that I understand this, I think correct thing to do is just to remove the call to updateImager from option().

johnhaddon replied:
Cool - thanks for the explanation.

This does make me think that maybe we should be avoiding the work of `updateDrivers()` when it isn't necessary though. We care about the performance of interactive updates for things like attributes and transforms, and doing the work of `updateDrivers()` unnecessarily for every render can't help with that. Maybe we need a flag to say if the outputs are dirty, so we can avoid the work?

danieldresser-ie replied:
I tested performance, and was unable to observe any noticeable slowdown in a worst case render ( a 10x10 pixel render with no shading, and 1000 aovs ). After talking to John, we're going to leave this as is for now.

danieldresser-ie commented on src/IECoreArnold/Renderer.cpp:
```diff
@@ -440,37 +440,35 @@ class ArnoldRendererBase : public IECoreScenePreview::Renderer
+const IECore::InternedString g_cameraOverrideInternedString( "cameraOverride" );
+const IECore::InternedString g_customAttributesInternedString( "customAttributes" );
+const IECore::InternedString g_driverNodeTypeInternedString( "driverNodeType" );
+const IECore::InternedString g_fileNameInternedString( "fileName" );
```
I'm a bit confused by this - this is used to set the file name on the driver ... that seems like a pretty regular parameter? Do you want me to prefix it with `__`? It seems like the only time that would really come up is if the user tried to add a custom attribute named "fileName" ... by prefixing the reserved name with `__`, I guess we could allow the user to set a user attribute name "fileName" independently of the actual filename, but I'm not sure that's desirable?

johnhaddon replied:
If it was a regular parameter, it would be called `filename` (all lowercase) to match the actual driver parameter name. My interpretation of a container called `driverParameters` would be that it contains name/value pairs corresponding directly to the parameters of the driver node. But we also using it to "smuggle" additional data into the ArnoldDriver constructor, and I thought that would be clearer if we prefix these "virtual" parameters with something. Maybe I'm wrong, but that was my thought process...

danieldresser-ie replied:
I'll try to give an impartial summary of our offline conversation: putting both arbitrary parameters, and parameters with special meaning in the same dictionary does bother us both a bit ( and John a bit more than me ), but neither of us came up with a solution we liked more. Putting everything in one bucket allows us to use generic code to report problems with display drivers with mismatched parameters, and because that error reporting code exposes the names to users, we don't want to use a weird prefix. So unless we come up with a better idea, the current plan is to leave this as is.

danieldresser-ie commented:
I think I've now addressed all feedback. The fixup PR seems fairly easy to review, so I haven't squashed it yet, but as far as I know, this should be good to go after squashing.

johnhaddon commented:
> I think I've now addressed all feedback.

I think you might have only pushed your changes to your fork, but not to the main repo (because this is an Arnold PR, it needed to be made from the main repo for CI to run). We also need to retarget to main still (the PR is targeting 1.5, and on your fork the Changes.md updates are under the 1.5 heading too). Could you sort that out before I take a final look please?

danieldresser-ie commented:
I'm a bit confused by this comment - I did forgot to update the branch on main. But your comment about the Changes.md makes it sound like you did actually look at the fixup commit on my branch - I put the update under 1.x.x.x with the other pending changes on main, I think it's correct?

johnhaddon commented:
> makes it sound like you did actually look at the fixup commit on my branch 

I did, but I misread the heading there - it is correctly under 1.x.x.x as you say. Sorry for the muddle.

danieldresser-ie commented:
Removed "__" prefix, squashed and merged.
